File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int[] e=new int[height.length];
        int[] f=new int[e.length];
        e[0]=height[0];
        for(int i=1;i<height.length;i++){
            e[i]=Math.max(e[i-1],height[i]);
        }
        f[height.length-1]=height[height.length-1];
        for(int i=height.length-2;i>=0;i--){
            f[i]=Math.max(f[i+1],height[i]);
        }
        int result=0;
        for(int i=0;i<height.length;i++){
            result+=Math.min(e[i],f[i])-height[i];
        }
        return result;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int n=0;
    vector<int> height;
    void find_left_max(vector<int> &left_max){
        left_max[0]=height[0];
        for(int e=1;e<n;e++){
            left_max[e]=max(height[e],left_max[e-1]);
        }
    }
    void find_right_max(vector<int> &right_max){
        right_max[n-1]=height[n-1];
        for(int e=n-2;e>=0;e--){
            right_max[e]=max(right_max[e+1],height[e]);
        }
    }
    int trap(vector<int>& h) {
        height=h;
        n=height.size();
        vector<int> left_max(n);
        vector<int> right_max(n);
        find_left_max(left_max);
        find_right_max(right_max);
        int result=0;
        for(int e=0;e<n;e++){
            result+=min(left_max[e],right_max[e])-height[e];
        }
        return result;
    }
};